### PR TITLE
[Android][iOS] Fix open page twitching when toolbar is hidden in page

### PR DESCRIPTION
**Pull request by rotorgames**
### Description of Change

If Page is set HasNavigationBar = false and open it after page which has HasNavigationBar = true then user see twitchings and it is bad.

Before:
toolbar-before

After: 
toolbar-after
### Bugs Fixed
- It is not bug, but it is not beautiful user experience
### API Changes

Add:
- Xamarin.Forms.Page.LayoutChild(VisualElement element, Rectangle area)

### Behavioral Changes

No changes
### PR Checklist
- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

**Comment by vurf** [comment]
+1

**Comment by BudyakovAlex** [comment]
+1

**Comment by Sky1arkk** [comment]
+1

**Comment by V0v1kkk** [comment]
+1

**Comment by StephaneDelcroix** [comment]
In the PR summary, you checked the box "Has tests" but I can't see any...

File: Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs [hunk 1]
```diff
@@ -300,17 +298,20 @@ protected override void OnLayout(bool changed, int l, int t, int r, int b)
...
 
-				if (ToolbarVisible)
+				if (isBar)
 				{
-					if (isBar)
+					if (ToolbarVisible)
 						bar.Layout(0, 0, r - l, barHeight);
 					else
-						child.Layout(0, barHeight + ContainerPadding, r, b);
+						bar.Layout(0, -1000, r, barHeight - 1000);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs]: what does 1000 means ?

Comment (rotorgames) [inline Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs]: @StephaneDelcroix It hides toolbar if `SetHideNavigationBar` is `false` in ContentPage.

`-1000` is not my idea, that was earlier. [NavigationPageRenderer.cs:313](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs#L313)

**Comment by antixaker** [comment]
+1

**Comment by rotorgames** [comment]
> In the PR summary, you checked the box "Has tests" but I can't see any...

@StephaneDelcroix I'm sorry. Please wait. I will create tests.

**Comment by StephaneDelcroix** [comment]
@rotorgames I was checking that you haven't forgot to commit them

**Comment by rotorgames** [comment]
@StephaneDelcroix I will finish soon and push commits. Please wait.

File: Xamarin.Forms.Core/NavigationPage.cs [hunk 2]
```diff
@@ -220,6 +222,14 @@ protected override bool OnBackButtonPressed()
 			return base.OnBackButtonPressed();
 		}
 
+		protected internal override void LayoutChild(VisualElement element, Rectangle area)
```
Comment (rotorgames) [inline Xamarin.Forms.Core/NavigationPage.cs]: Maybe do it public?

**Comment by rotorgames** [comment]
@StephaneDelcroix I added more changes and added tests. See it please.

File: Xamarin.Forms.Core/Page.cs [hunk 3]
```diff
@@ -175,15 +175,21 @@ protected virtual void LayoutChildren(double x, double y, double width, double h
...
 				}
 				else
 				{
-					Forms.Layout.LayoutChildIntoBoundingRegion(child, area);
+					LayoutChild(child, area);
 				}
 			}
 		}
 
+		protected internal virtual void LayoutChild(VisualElement element, Rectangle area)
```
Comment (rotorgames) [inline Xamarin.Forms.Core/Page.cs]: Maybe do it "protected" only (is not internal)?

**Comment by rotorgames** [comment]
@StephaneDelcroix P.S. iOS has same problem. If push new page in which toolbar is hided that all pages in stack will be resized.

Example (bug in Xamarin.Forms is not in my code):
toolbar-ios

**Comment by rotorgames** [comment]
@StephaneDelcroix I fixed a bug in iOS and optimized `ForceLayout` method in Android. `ForceLayout` is invoked once now. 

P.S. Can you test it on WP, UWP?

**Comment by Ptytz** [comment]
+1

**Comment by stanbav** [comment]
+1

**Comment by StephaneDelcroix** [comment]
@vurf, @BudyakovAlex, @Sky1arkk, @V0v1kkk, @antixaker, @Ptytz, @stanbav 
Have you seen 👍 at the bottom of the PR ? That's where you should register your votes.

All the rest adds noise.

Thanks for your understanding.

**Comment by rotorgames** [comment]
@StephaneDelcroix Any news? Do you see my changes in code?

**Comment by pauldipietro** [comment]
@rotorgames The prior iOS commit doesn't appear to entirely resolve the issue. As it exists, it looks like it's getting rid of the blank space -- which it is -- but the subsequently pushed page is not correctly sized. Refer to #370, but here is a visual comparison; first of those changes and then using the reproduction in the aforementioned PR:

**Comment by rotorgames** [comment]
@pauldipietro I pull from your branch (fix-bugzilla43955) and merge with my branch (rotorgames:fix-android-toolbar). I remove [NavigationRenderer.cs:319-344](https://github.com/xamarin/Xamarin.Forms/pull/370/files#diff-efa21a065745b4e261efc4bf6a5d8784R319) code.

All worked! My code fix all problems. My code works is correct

Also, I tested your "Bugzilla43995" test with my (rotorgames:fix-android-toolbar) and without #370. All worked is correct.

You can test it with #370 https://github.com/rotorgames/Xamarin.Forms/tree/fix-bugzilla43955
or without #370 https://github.com/rotorgames/Xamarin.Forms/tree/fix-android-toolbar-without

toolbar-compare

**Comment by rotorgames** [comment]
@StephaneDelcroix @pauldipietro Hi friends. I added many changes. All bugs are fixed.

Please, look to video:

Before:
[Before](https://drive.google.com/file/d/0B1w5ePBNpSdMTHlsbHd3WGVUcDQ/view?usp=sharing)

After:
[After](https://drive.google.com/file/d/0B1w5ePBNpSdMdnlIUWJGTGlZMjg/view?usp=sharing)

**Comment by rotorgames** [comment]
@StephaneDelcroix @pauldipietro Any news?

**Comment by rotorgames** [comment]
@StephaneDelcroix @pauldipietro Hello. You reviewed this PR? My friends and I want see this changes in release so much. Please, tell me, when you check this PR? Thank you.

**Comment by rotorgames** [comment]
@StephaneDelcroix @pauldipietro Hello. Will you review this PR? I created this PR four months ago and it was not added to release. Do you have problems with changes? I can help if need.

File: Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs [hunk 4]
```diff
@@ -382,6 +377,8 @@ void BarOnNavigationClick(object sender, AToolbar.NavigationClickEventArgs navig
 
 		void CurrentOnPropertyChanged(object sender, PropertyChangedEventArgs e)
 		{
+			if(e.PropertyName == NavigationPage.InternalPaddingProperty.PropertyName)
+				Element.ForceLayout();
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs]: isn't RequestLayout sufficient here ?

File: Xamarin.Forms.Platform.iOS/Renderers/NavigationRenderer.cs [hunk 5]
```diff
@@ -132,19 +151,16 @@ public override void ViewDidDisappear(bool animated)
...
+			var modelSize = _queuedSize.IsZero ? Element.Bounds.Size : _queuedSize;
 
-			var toolbar = _secondaryToolbar;
-			// Use 0 if the NavBar is hidden or will be hidden
-			var toolbarY = NavigationBarHidden || !NavigationPage.GetHasNavigationBar(Current) ? 0 : navBarFrame.Bottom;
-			toolbar.Frame = new RectangleF(0, toolbarY, View.Frame.Width, toolbar.Frame.Height);
+			ViewControllers
+				.Cast<ParentingViewController>()
+				.Select(vc => vc.Child)
+				.ForEach(UpdateInternalPadding);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.iOS/Renderers/NavigationRenderer.cs]: please use a foreach loop:
```
foreach (ParentingViewController vc in ViewControllers) 
    UpdateInternalPadding(vc.Child);
```

File: Xamarin.Forms.Platform.iOS/Renderers/NavigationRenderer.cs [hunk 6]
```diff
@@ -580,6 +612,20 @@ void UpdateBarTextColor()
 			}
 		}
 
+		void UpdateToolbarColors()
+		{
+			var barTextColor = ((NavigationPage)Element).BarTextColor;
+			var barBackgroundColor = ((NavigationPage)Element).BarBackgroundColor;
+
+			ViewControllers
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.iOS/Renderers/NavigationRenderer.cs]: foreach loop

File: Xamarin.Forms.Platform.iOS/Renderers/NavigationRenderer.cs [hunk 7]
```diff
@@ -37,7 +38,25 @@ public NavigationRenderer()
 			
 		}
 
-		Page Current { get; set; }
+		private Page Current
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.iOS/Renderers/NavigationRenderer.cs]: you don't have to specify the private modifier, we all know that it is private

**Comment by rotorgames** [comment]
@StephaneDelcroix Ok. I will do it soon.

File: Xamarin.Forms.Core/NavigationPage.cs [hunk 8]
```diff
@@ -184,6 +186,12 @@ public Task PushAsync(Page page)
 
 		public event EventHandler<NavigationEventArgs> Pushed;
 
+		internal static void SetInternalPadding(Page page, Thickness bound)
+		{
+			if(!((Thickness)page.GetValue(InternalPaddingProperty)).Equals(bound))
+				page.SetValue(InternalPaddingProperty, bound);
```
Comment (rotorgames) [inline Xamarin.Forms.Core/NavigationPage.cs]: @StephaneDelcroix Do you want to add this in the `validateValue' delegate of the BP?

```chsarp
!((Thickness)page.GetValue(InternalPaddingProperty)).Equals(bound)
```

If yes then BP will throw exception if old value equals to new value but this is not error. Just I don't want set new value if it equals old value. Maybe do you know better solution?

**Comment by rotorgames** [comment]
@StephaneDelcroix 

> InternalPadding needs a better name, like ToolbarPadding and a note in the code telling what is purpose is "this is the padding we use while transitioning from page with different toolbars to avoid the jump of content"

But the InternalPadding can be used for many solutions not only for the toolbar padding. In future we can use it not only for the toolbar, but the name "ToolbarPadding" will mislead.

Maybe can I keep the `InternalPadding` name or will you give the better name?

File: Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs (hunk 4 above)
Comment (rotorgames) [inline Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs]: @StephaneDelcroix There is not RequestLayout in NavigationPage, only ForceLayout

**Comment by rotorgames** [comment]
@StephaneDelcroix I fixed several remarks. I'm waiting for your answer.

**Comment by rotorgames** [comment]
@StephaneDelcroix Hi. Any news about it?

**Comment by rotorgames** [comment]
@StephaneDelcroix @pauldipietro @jassmith Hi. Any news about it? I created this PR more then half of year ago. This is a serious problem. Can you say about it anything?

**Comment by michaelstonis** [comment]
Any movement on this bug? This has been sitting out here for some time without comment.

**Comment by samhouts** [comment]
We appreciate all of the hard work that went into this PR! We are looking at an alternate fix in #962. Please direct your comments over there. We will go ahead and close this request, but thank you for the effort and research you did here!

**Pull request closed without merging**